private/bufpkg/bufmodule/bufmodulestore/module_data_store.go

```diff
@@ -317,11 +313,95 @@ func (p *moduleDataStore) putModuleData(
 			}
 		}()
 	} else {
-		moduleCacheBucket, err = p.getReadWriteBucketForDir(moduleKey)
+		dirPath, err := getModuleDataStoreDirPath(moduleKey)
 		if err != nil {
 			return err
 		}
+		p.logDebugModuleKey(
+			moduleKey,
+			"module data store dir read write bucket",
+			zap.String("dirPath", dirPath),
+		)
+		moduleCacheBucket = storage.MapReadWriteBucket(p.bucket, storage.MapOnPrefix(dirPath))
+
+		// Only attempt to get a file locks when storing individual files.
+		// Before writing to the module directory, first get a shared lock and check module.yaml
+		var readUnlocker filelock.Unlocker
+		readUnlocker, err = p.filelocker.RLock(ctx, dirPath+externalModuleDataLockFileExt)
+		if err != nil {
+			p.logger.Debug("failed to get rlock for putModuleData", zap.Error(err))
+			return err
+		}
+		defer func() {
+			if readUnlocker != nil {
+				if err := readUnlocker.Unlock(); err != nil {
+					p.logger.Debug("failed to runlock for putModuleData", zap.Error(err))
+					retErr = multierr.Append(retErr, err)
+				}
+			}
+		}()
+		data, err := storage.ReadPath(ctx, moduleCacheBucket, externalModuleDataFileName)
+		p.logDebugModuleKey(
+			moduleKey,
+			fmt.Sprintf("module data store put read check %s", externalModuleDataFileName),
```

Comment:
Was this copied from previous code? I don't understand what this is, perhaps it was my fault though.

Reply:
This was not -- this is part of the comment above:

```
		// Before writing to the module directory, first get a shared lock and check module.yaml
```

We are reading the `module.yaml` preemptively before attempting to write new data to the cache.